rclc/include/rclc/executor.h

```diff
@@ -0,0 +1,387 @@
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ *
+ * \param [inout] executor pointer to initialized executor
+ * \param[in] timeout_ns  timeout in millisonds
+ * \return `RCL_RET_OK` if spin_once operation was successful
+ * \return `RCL_RET_INVALID_ARGUMENT` if any parameter is a null pointer
+ * \return `RCL_RET_TIMEOUT` if rcl_wait() returned timeout (aka no data is avaiable during until the timeout)
+ * \return `RCL_RET_ERROR` if any other error occured
+ */
+rcl_ret_t
+rclc_executor_spin_some(
+  rclc_executor_t * executor,
+  const uint64_t timeout_ns);
+
+/**
+ *  The spin function checks for new data at DDS queue as long as ros context is available.
+ *  It calls {@link rclc_executor_spin_some()} as long as rcl_is_context_is_valid() returns true.
+ *
+ *  Memory is dynamically allocated within rcl-layer, when DDS queue is accessed with rcl_wait_set_init()
+ *  (in spin_some function)
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ *
+ * \param [inout] executor pointer to initialized executor
+ * \return `RCL_RET_OK` if spin operation was successful
+ * \return `RCL_RET_INVALID_ARGUMENT` if executor is a null pointer
+ * \return `RCL_RET_ERROR` if any other error occured
+ */
+rcl_ret_t
+rclc_executor_spin(rclc_executor_t * executor);
+
+
+/**
+ *  The spin_period function checks for new data at DDS queue as long as ros context is available.
+ *  It is called every period nanoseconds.
+ *  It calls {@link rclc_executor_spin_some()} as long as rcl_is_context_is_valid() returns true.
+ *
+ *  Memory is dynamically allocated within rcl-layer, when DDS queue is accessed with rcl_wait_set_init()
+ *  (in spin_some function)
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ *
+ * \param [inout] executor pointer to initialized executor
+ * \param [in] period in nanoseconds
+ * \return `RCL_RET_OK` if spin operation was successful
+ * \return `RCL_RET_INVALID_ARGUMENT` if executor is a null pointer
+ * \return `RCL_RET_ERROR` if any other error occured
+ */
+rcl_ret_t
+rclc_executor_spin_period(
+  rclc_executor_t * executor,
+  const uint64_t period);
+
+
+/*
+ The reason for splitting this function up, is to be able to write a unit test.
+ The spin_period is an endless loop, therefore it is not possible to stop after x iterations. The function
+ rclc_executor_spin_period_ implements one iteration and the function
+ rclc_executor_spin_period implements the endless while-loop. The unit test covers only
+ rclc_executor_spin_period_.
+*/
```

Comment:
I am not sure if the style (different from the style above) will be picked up from doxygen like this.

Reply:
same below

Reply:
fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b